blog/content/post/2023-06-07-twpm-spi-fix.md

```diff
@@ -0,0 +1,855 @@
+	.frequency = 25000000,
+	.operation = SPI_WORD_SET(8) | SPI_TRANSFER_MSB | SPI_OP_MODE_SLAVE,
+};
+
+const uint8_t testdata[32] = {
+	0x3d, 0x7f, 0x85, 0xc3, 0x86, 0x2f, 0x14, 0x2a, 0xa2, 0x67, 0x1d, 0xd7, 0xfa, 0xa8, 0x3a, 0x42,
+	0xf8, 0x12, 0xd2, 0xa1, 0x04, 0xcc, 0xe2, 0xc6, 0x78, 0x73, 0x09, 0xe6, 0xd8, 0xc5, 0x0e, 0xba,
+};
+
+void main(void) {
+	spi_dev = device_get_binding(DEVICE_DT_NAME(DT_NODELABEL(spi2)));
+	if (!device_is_ready(spi_dev)) {
+		printk("SPI controller is not ready!\n");
+		return;
+	}
+
+	struct spi_buf spi_tx_buf = {
+		.buf = (void *)testdata,
+		.len = sizeof testdata,
+	};
+
+	struct spi_buf_set spi_tx_buf_set = {
+		.buffers = &spi_tx_buf,
+		.count = 1,
+	};
+
+	while (true) {
+		int ret = spi_write(spi_dev, &spi_cfg, &spi_tx_buf_set);
+		if (ret < 0)
+			printk("spi_write failed: %d\n", ret);
+		else
+			printk("spi transfer complete\n");
+	}
+}
+```
+
+It is necessary to enable some configs:
+
+```shell
+CONFIG_GPIO=y
+CONFIG_SPI=y
+CONFIG_SPI_SLAVE=y
+CONFIG_SPI_STM32_DMA=y
+```
+
+And configure DMA channels:
+
+```shell
+&dma1 {
+	status = "okay";
+};
+
+&spi1 {
+	status = "disabled";
+};
+
+&spi2 {
+	/*
+	 * See https://docs.zephyrproject.org/3.0.0/reference/devicetree/bindings/dma/st%2Cstm32-dma-v2.html
+	 */
+	dmas = <&dma1 5 1 0x20440>,
+	       <&dma1 4 1 0x20480>;
+	dma-names = "tx", "rx";
+};
+
+&spi3 {
+	status = "disabled";
+};
+```
+
+If we run this code we will see SPI errors running down the console:
+
+```shell
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+spi_write failed: -11
+```
+
+This is the first problem with Zephyr's SPI driver - each transfer has a
+one-second timeout. While this may be desirable behavior for SPI master (it
+could be used for error recovery, for example, to power cycle the slave if it
+doesn't respond), it breaks SPI slave. The slave must be ready to give a
+response when the transfer commences - appropriate data must already be loaded
+in FIFO. Here we get stuck in an endless loop, queuing the transfer, aborting
+it, and queuing it again.
+
+The problem can be worked around by patching the `wait_dma_rx_tx_done` function
+in `spi_ll_stm32.c`. The original function looks like this:
+
+```c
+static int wait_dma_rx_tx_done(const struct device *dev)
+{
+	struct spi_stm32_data *data = dev->data;
+	int res = -1;
+
+	while (1) {
+		res = k_sem_take(&data->status_sem, K_MSEC(1000));
+		if (res != 0) {
+			return res;
+		}
+...
+```
+
+Just replace `K_MSEC(1000)` with `K_FOREVER`.
+
+Now running `spitest` at 100 KHz yields the following result:
+
+![STM32 SPI slave at 100 KHz](/img/stm32-spislave-100khz.png)
+
+The transfer works properly at 100 KHz. At 10 MHz the transfer sometimes works,
+sometimes does not:
+
+![STM32 SPI slave at 10 MHz](/img/stm32-spislave-10mhz.png)
+
+At 24 MHz transfer is completely corrupted. We have been looking for a solution
+in Zephyr Issues and Pull Requests but found nothing useful.
```

Comment:
Please provide links for both

Reply:
@macpijan I found nothing useful there so there is no link.